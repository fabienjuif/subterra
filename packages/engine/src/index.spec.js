import { createPatch } from 'diff'
import { createEngine } from './index'
import {
  cards,
  tiles,
  archetypes,
  EndCard,
  GazCard,
  WaterCard,
  ShakeCard,
  HorrorCard,
} from '@subterra/data'

expect.addSnapshotSerializer({
  test: () => true,
  print: (str) => str,
})

describe('engine without mock', () => {
  it('should be replayable', () => {
    const play = () => {
      const engine = createEngine()

      engine.dispatch({
        type: '@cards>init',
        payload: [cards[2], cards[3], cards[1], cards[0]],
      })
      engine.dispatch({
        type: '@tiles>init',
        payload: [tiles[0], tiles[2], tiles[3], tiles[1]],
      })
      engine.dispatch({
        type: '@players>init',
        payload: [archetypes[0], archetypes[1], archetypes[2]].map(
          (archetype) => ({
            ...archetype,
            archetype,
            name: `xXx-${archetype.type}-xXx`,
          }),
        ),
      })
      engine.dispatch('@cards>pick')
      // TODO: here play a fake game

      return engine
    }

    // 3 times should be enough
    const firstState = play().getState()
    expect(play().getState()).toEqual(firstState)
    expect(play().getState()).toEqual(firstState)
  })

  it('should connect behaviours to actions', () => {
    const engine = createEngine()

    const dispatchAndSnap = (action) => {
      const previousState = engine.getState()
      engine.dispatch(action)
      const afterState = engine.getState()

      expect(
        createPatch(
          typeof action === 'object' ? action.type : action,
          JSON.stringify(previousState, null, 2),
          JSON.stringify(afterState, null, 2),
        ),
      ).toMatchSnapshot()
    }

    dispatchAndSnap({
      type: '@cards>init',
      payload: [ShakeCard, GazCard, WaterCard, HorrorCard, EndCard],
    })
    dispatchAndSnap({
      type: '@players>init',
      payload: [archetypes[0], archetypes[1], archetypes[2]].map(
        (archetype) => ({
          ...archetype,
          archetype,
          name: `xXx-${archetype.type}-xXx`,
        }),
      ),
    })
    dispatchAndSnap({ type: '@dices>init', payload: [2, 3, 4, 5] })
    dispatchAndSnap('@cards>pick')

    // TODO: complete this after we can discover and move (or explore)
  })
})
