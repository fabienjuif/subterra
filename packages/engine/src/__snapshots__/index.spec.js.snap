// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine without mock should connect behaviours to actions 1`] = `
Index: @cards>init
===================================================================
--- @cards>init
+++ @cards>init
@@ -7,9 +7,33 @@
     "deck": []
   },
   "players": [],
   "enemies": [],
-  "deckCards": [],
+  "deckCards": [
+    {
+      "type": "shake",
+      "damage": 1,
+      "duration": 1
+    },
+    {
+      "type": "gaz",
+      "damage": 2,
+      "duration": 1
+    },
+    {
+      "type": "water",
+      "damage": 2,
+      "duration": 1
+    },
+    {
+      "type": "enemy",
+      "damage": 3,
+      "duration": 1
+    },
+    {
+      "type": "end"
+    }
+  ],
   "activeCard": {},
   "grid": [
     {
       "id": 0,
@@ -26,7 +50,36 @@
   "playerActions": {
     "possibilities": []
   },
   "technical": {
-    "actions": []
+    "actions": [
+      {
+        "type": "@cards>init",
+        "payload": [
+          {
+            "type": "shake",
+            "damage": 1,
+            "duration": 1
+          },
+          {
+            "type": "gaz",
+            "damage": 2,
+            "duration": 1
+          },
+          {
+            "type": "water",
+            "damage": 2,
+            "duration": 1
+          },
+          {
+            "type": "enemy",
+            "damage": 3,
+            "duration": 1
+          },
+          {
+            "type": "end"
+          }
+        ]
+      }
+    ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 2`] = `
Index: @players>init
===================================================================
--- @players>init
+++ @players>init
@@ -5,9 +5,141 @@
   "tiles": {
     "remaining": 0,
     "deck": []
   },
-  "players": [],
+  "players": [
+    {
+      "type": "medic",
+      "health": 3,
+      "strength": 6,
+      "skills": [
+        {
+          "type": "heal",
+          "perGame": null,
+          "perTurn": null,
+          "cost": 1
+        },
+        {
+          "type": "sprint",
+          "perGame": null,
+          "perTurn": null,
+          "cost": 1
+        }
+      ],
+      "archetype": {
+        "type": "medic",
+        "health": 3,
+        "strength": 6,
+        "skills": [
+          {
+            "type": "heal",
+            "perGame": null,
+            "perTurn": null,
+            "cost": 1
+          },
+          {
+            "type": "sprint",
+            "perGame": null,
+            "perTurn": null,
+            "cost": 1
+          }
+        ]
+      },
+      "name": "xXx-medic-xXx",
+      "id": "medic",
+      "x": 0,
+      "y": 0,
+      "actionPoints": 2,
+      "current": true,
+      "first": true
+    },
+    {
+      "type": "geologist",
+      "health": 3,
+      "strength": 3,
+      "skills": [
+        {
+          "type": "intuition",
+          "perGame": null,
+          "perTurn": null,
+          "cost": 0
+        },
+        {
+          "type": "clear",
+          "perGame": null,
+          "perTurn": null,
+          "cost": 1
+        }
+      ],
+      "archetype": {
+        "type": "geologist",
+        "health": 3,
+        "strength": 3,
+        "skills": [
+          {
+            "type": "intuition",
+            "perGame": null,
+            "perTurn": null,
+            "cost": 0
+          },
+          {
+            "type": "clear",
+            "perGame": null,
+            "perTurn": null,
+            "cost": 1
+          }
+        ]
+      },
+      "name": "xXx-geologist-xXx",
+      "id": "geologist",
+      "x": 0,
+      "y": 0,
+      "actionPoints": 2
+    },
+    {
+      "type": "scoot",
+      "health": 3,
+      "strength": 2,
+      "skills": [
+        {
+          "type": "guide",
+          "cost": 0,
+          "perGame": 3,
+          "perTurn": null
+        },
+        {
+          "type": "furtivity",
+          "cost": 0,
+          "perGame": null,
+          "perTurn": null
+        }
+      ],
+      "archetype": {
+        "type": "scoot",
+        "health": 3,
+        "strength": 2,
+        "skills": [
+          {
+            "type": "guide",
+            "cost": 0,
+            "perGame": 3,
+            "perTurn": null
+          },
+          {
+            "type": "furtivity",
+            "cost": 0,
+            "perGame": null,
+            "perTurn": null
+          }
+        ]
+      },
+      "name": "xXx-scoot-xXx",
+      "id": "scoot",
+      "x": 0,
+      "y": 0,
+      "actionPoints": 2
+    }
+  ],
   "enemies": [],
   "deckCards": [
     {
       "type": "shake",
@@ -47,9 +179,46 @@
       "status": []
     }
   ],
   "playerActions": {
-    "possibilities": []
+    "possibilities": [
+      {
+        "type": "@players>look",
+        "payload": {
+          "playerId": "medic",
+          "x": 0,
+          "y": -1,
+          "cost": 1
+        }
+      },
+      {
+        "type": "@players>look",
+        "payload": {
+          "playerId": "medic",
+          "x": -1,
+          "y": 0,
+          "cost": 1
+        }
+      },
+      {
+        "type": "@players>look",
+        "payload": {
+          "playerId": "medic",
+          "x": 1,
+          "y": 0,
+          "cost": 1
+        }
+      },
+      {
+        "type": "@players>look",
+        "payload": {
+          "playerId": "medic",
+          "x": 0,
+          "y": 1,
+          "cost": 1
+        }
+      }
+    ]
   },
   "technical": {
     "actions": [
       {
@@ -78,8 +247,130 @@
           {
             "type": "end"
           }
         ]
+      },
+      {
+        "type": "@players>init",
+        "payload": [
+          {
+            "type": "medic",
+            "health": 3,
+            "strength": 6,
+            "skills": [
+              {
+                "type": "heal",
+                "perGame": null,
+                "perTurn": null,
+                "cost": 1
+              },
+              {
+                "type": "sprint",
+                "perGame": null,
+                "perTurn": null,
+                "cost": 1
+              }
+            ],
+            "archetype": {
+              "type": "medic",
+              "health": 3,
+              "strength": 6,
+              "skills": [
+                {
+                  "type": "heal",
+                  "perGame": null,
+                  "perTurn": null,
+                  "cost": 1
+                },
+                {
+                  "type": "sprint",
+                  "perGame": null,
+                  "perTurn": null,
+                  "cost": 1
+                }
+              ]
+            },
+            "name": "xXx-medic-xXx"
+          },
+          {
+            "type": "geologist",
+            "health": 3,
+            "strength": 3,
+            "skills": [
+              {
+                "type": "intuition",
+                "perGame": null,
+                "perTurn": null,
+                "cost": 0
+              },
+              {
+                "type": "clear",
+                "perGame": null,
+                "perTurn": null,
+                "cost": 1
+              }
+            ],
+            "archetype": {
+              "type": "geologist",
+              "health": 3,
+              "strength": 3,
+              "skills": [
+                {
+                  "type": "intuition",
+                  "perGame": null,
+                  "perTurn": null,
+                  "cost": 0
+                },
+                {
+                  "type": "clear",
+                  "perGame": null,
+                  "perTurn": null,
+                  "cost": 1
+                }
+              ]
+            },
+            "name": "xXx-geologist-xXx"
+          },
+          {
+            "type": "scoot",
+            "health": 3,
+            "strength": 2,
+            "skills": [
+              {
+                "type": "guide",
+                "cost": 0,
+                "perGame": 3,
+                "perTurn": null
+              },
+              {
+                "type": "furtivity",
+                "cost": 0,
+                "perGame": null,
+                "perTurn": null
+              }
+            ],
+            "archetype": {
+              "type": "scoot",
+              "health": 3,
+              "strength": 2,
+              "skills": [
+                {
+                  "type": "guide",
+                  "cost": 0,
+                  "perGame": 3,
+                  "perTurn": null
+                },
+                {
+                  "type": "furtivity",
+                  "cost": 0,
+                  "perGame": null,
+                  "perTurn": null
+                }
+              ]
+            },
+            "name": "xXx-scoot-xXx"
+          }
+        ]
       }
     ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 3`] = `
Index: @seeds>init
===================================================================
--- @seeds>init
+++ @seeds>init
@@ -1,7 +1,14 @@
 {
   "seeds": {
-    "public": false
+    "master": "87FZ2jefz",
+    "private": true,
+    "dices": "7M8cMGMdlm",
+    "dicesNext": "7M8cMGMdlm",
+    "cards": "Y2laJE1czq",
+    "cardsNext": "Y2laJE1czq",
+    "tiles": "5hk7Zpp7wP",
+    "tilesNext": "5hk7Zpp7wP"
   },
   "tiles": {
     "remaining": 0,
     "deck": []
@@ -369,8 +376,14 @@
             },
             "name": "xXx-scoot-xXx"
           }
         ]
+      },
+      {
+        "type": "@seeds>init",
+        "payload": {
+          "master": "87FZ2jefz"
+        }
       }
     ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 4`] = `
Index: @cards>pick
===================================================================
--- @cards>pick
+++ @cards>pick
@@ -2,9 +2,9 @@
   "seeds": {
     "master": "87FZ2jefz",
     "private": true,
     "dices": "7M8cMGMdlm",
-    "dicesNext": "7M8cMGMdlm",
+    "dicesNext": "7M8cMGMdlm@@0.04602145714249605",
     "cards": "Y2laJE1czq",
     "cardsNext": "Y2laJE1czq",
     "tiles": "5hk7Zpp7wP",
     "tilesNext": "5hk7Zpp7wP"
@@ -15,9 +15,9 @@
   },
   "players": [
     {
       "type": "medic",
-      "health": 3,
+      "health": 2,
       "strength": 6,
       "skills": [
         {
           "type": "heal",
@@ -60,9 +60,9 @@
       "first": true
     },
     {
       "type": "geologist",
-      "health": 3,
+      "health": 2,
       "strength": 3,
       "skills": [
         {
           "type": "intuition",
@@ -103,9 +103,9 @@
       "actionPoints": 2
     },
     {
       "type": "scoot",
-      "health": 3,
+      "health": 2,
       "strength": 2,
       "skills": [
         {
           "type": "guide",
@@ -148,13 +148,8 @@
   ],
   "enemies": [],
   "deckCards": [
     {
-      "type": "shake",
-      "damage": 1,
-      "duration": 1
-    },
-    {
       "type": "gaz",
       "damage": 2,
       "duration": 1
     },
@@ -171,9 +166,13 @@
     {
       "type": "end"
     }
   ],
-  "activeCard": {},
+  "activeCard": {
+    "type": "shake",
+    "damage": 1,
+    "duration": 1
+  },
   "grid": [
     {
       "id": 0,
       "type": "start",
@@ -382,8 +381,192 @@
         "type": "@seeds>init",
         "payload": {
           "master": "87FZ2jefz"
         }
+      },
+      {
+        "type": "@cards>pick"
+      },
+      {
+        "type": "@cards>shake",
+        "payload": {
+          "card": {
+            "type": "shake",
+            "damage": 1,
+            "duration": 1
+          }
+        }
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerId": "medic",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "medic"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerId": "geologist",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "geologist"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerId": "scoot",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "scoot"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerId": "medic",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "medic"
+            }
+          },
+          "value": 2
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerId": "geologist",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "geologist"
+            }
+          },
+          "value": 1
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerId": "scoot",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerId": "scoot"
+            }
+          },
+          "value": 1
+        }
+      },
+      {
+        "type": "@players>damage",
+        "payload": {
+          "damage": 1,
+          "from": {
+            "card": {
+              "type": "shake",
+              "damage": 1,
+              "duration": 1
+            }
+          },
+          "playerId": "medic"
+        }
+      },
+      {
+        "type": "@players>damage",
+        "payload": {
+          "damage": 1,
+          "from": {
+            "card": {
+              "type": "shake",
+              "damage": 1,
+              "duration": 1
+            }
+          },
+          "playerId": "geologist"
+        }
+      },
+      {
+        "type": "@players>damage",
+        "payload": {
+          "damage": 1,
+          "from": {
+            "card": {
+              "type": "shake",
+              "damage": 1,
+              "duration": 1
+            }
+          },
+          "playerId": "scoot"
+        }
       }
     ]
   }
 }
\\ No newline at end of file

`;
